gravitee-am-factor/gravitee-am-factor-recovery-code/src/main/java/io/gravitee/am/factor/recovery/code/provider/RecoveryCodeFactorProvider.java

```diff
@@ -108,17 +108,20 @@ public Completable generateRecoveryCode(FactorContext context) {
         return addRecoveryCodeFactor(context, enrolledFactor);
     }
 
-    private Completable addRecoveryCodeFactor(FactorContext context, EnrolledFactor enrolledFactor) {
-        try {
-            final UserService userService = context.getComponent(UserService.class);
-            enrolledFactor.setSecurity(createEnrolledFactorSecurity());
-
-            return userService
-                    .addFactor(context.getUser().getId(), enrolledFactor, new DefaultUser(context.getUser()))
-                    .ignoreElement();
-        } catch (Exception ex) {
-            return Completable.error(ex);
-        }
+    private Single<EnrolledFactorSecurity> addRecoveryCodeFactor(FactorContext context, EnrolledFactor enrolledFactor) {
+        return Single.create(emitter -> {
+            try {
+                final UserService userService = context.getComponent(UserService.class);
+                final EnrolledFactorSecurity enrolledFactorSecurity = createEnrolledFactorSecurity();
+                enrolledFactor.setSecurity(enrolledFactorSecurity);
+
+                userService
+                        .addFactor(context.getUser().getId(), enrolledFactor, new DefaultUser(context.getUser()))
+                        .subscribe(user -> emitter.onSuccess(enrolledFactorSecurity));
+            } catch (Exception exception) {
+                emitter.onError(exception);
+            }
+        });
```

Comment:
No need to create a Single here, just do : 

```
try {
  final UserService userService = context.getComponent(UserService.class);
  final EnrolledFactorSecurity enrolledFactorSecurity = createEnrolledFactorSecurity();
  enrolledFactor.setSecurity(enrolledFactorSecurity);

  return userService
    .addFactor(context.getUser().getId(), enrolledFactor, new DefaultUser(context.getUser()));
} catch (Exception exception) {
  return Single.error(exception);
}
```

Reply:
updated as suggested